File: problems/golang/11.binary-tree/main/is_binary_search_tree_test.go
package main

import (
	"github.com/AmirAhmadzadeh/problems/binary_tree"
	"testing"
)

// fisrt test case
func TestIsBinarySearchTree(t *testing.T) {
	binaryTree := binary_tree.New(12)
	root := binaryTree.GetRoot()
	generatedNode := root.PushLeftLeaf(7)
	generatedNode.PushLeftLeaf(5)
	generatedNode.PushRightLeaf(9)

	generatedNode = root.PushRightLeaf(18)
	generatedNode.PushLeftLeaf(16)
	generatedNode.PushRightLeaf(25)

	result := isBinarySearchTree(root)

	// result should be true
	if !result {
		t.Errorf("1.Expected (%v) is not same as"+
			" actual string (%v)", true, result)
	}
}

func TestIsBinarySearchTree2(t *testing.T) {
	binaryTree := binary_tree.New(13)
	root := binaryTree.GetRoot()

	generatedNode := root.PushLeftLeaf(6)
	generatedNode.PushLeftLeaf(2)

	generatedNode = root.PushRightLeaf(17)
	generatedNode.PushLeftLeaf(10)
	generatedNode.PushRightLeaf(22)

	result := isBinarySearchTree(root)

	// result should be false
	if result {
		t.Errorf("1.Expected (%v) is not same as"+
			" actual string (%v)", false, result)
	}
}

File: problems/ts/11.binary-tree/isBinarySearchTree.test.ts
import { describe, test, assert } from 'vitest';
import {
    BinaryTree,
    BinaryTreeImpl,
    Node,
    NodeImpl,
    isBinarySearchTree,
} from './';

describe('binary tree: problem#4', () => {
    let binaryTree: BinaryTree<number>;
    let root: Node<number>;

    test('', () => {
        binaryTree = new BinaryTreeImpl<number>(new NodeImpl(12));
        root = binaryTree.getRoot() as Node<number>;
        let generatedNode = root.pushLeftLeaf(7);
        generatedNode.pushLeftLeaf(5);
        generatedNode.pushRightLeaf(9);

        generatedNode = root.pushRightLeaf(18);
        generatedNode.pushLeftLeaf(16);
        generatedNode.pushRightLeaf(25);

        assert.equal(isBinarySearchTree({ ...root }), true);
    });

    test('should check whether the tree is bst or not', () => {
        binaryTree = new BinaryTreeImpl<number>(new NodeImpl(13));
        root = binaryTree.getRoot() as Node<number>;
        let generatedNode = root.pushLeftLeaf(6);
        generatedNode.pushLeftLeaf(2);

        generatedNode = root.pushRightLeaf(17);
        generatedNode.pushLeftLeaf(10);
        generatedNode.pushRightLeaf(22);

        assert.equal(isBinarySearchTree({ ...root }), false);
    });
});
